#19 - Refactor the code and add tests

bjtulinl opened the pull request:
- Restructure the code
- Add the unit test to BIO encoding. (A test file which contains the BIO labels will be loaded in the INCEpTION project and the test function and then the test function will compare the labeled sequences from the two parts and determine if they are the same.)

bjtulinl committed:
#19 - Refactor the code and add tests

- Restructure the code
- Add the unit test to BIO encoding. (A test file which contains the BIO labels will be loaded in the INCEpTION project and the test function and then the test function will compare the labeled sequences from the two parts and determine if they are the same.)

reckart commented on app_flair.py:
```diff
@@ -71,11 +81,11 @@ def route_train_ner():
+        t1 = threading.Thread(target=flair_train_ner, args=(train, dev))
         t1.start()
-        return ('', 204)
+        return ("", 204)
```
What is the empty string? I assume the number is a HTTP code. The could would be more readable if constants were used instead of "magic numbers".

reckart commented on app_flair.py:
```diff
@@ -116,126 +126,185 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+    tokens = []
+    flag = False
+    while i < token_list_len and j < ner_list_len:
+        if ner_list[j].value is None or "S-" + ner_list[j].value not in tagset:
```
What is "S-"?

bjtulinl replied:
"S-" is the single ner label.

reckart commented on app_flair.py:
```diff
@@ -116,126 +126,185 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+                token_list[i].begin > ner_list[j].begin
+                and token_list[i].end == ner_list[j].end
+        ):
+            token.add_tag("ner", "E-" + ner_list[j].value)
```
What is "E-"?

bjtulinl replied:
"E-" is the end of a multiword ner label.

reckart commented on app_flair.py:
```diff
@@ -116,126 +126,185 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+        token_list = list(cas.select_covered(TOKEN_TYPE, sentence))
+        if len(pos_list) != len(token_list):
+            continue
+        flag = False
```
What does "flag" indicate? There is no comment explaining it and the variable name could mean anything.

bjtulinl replied:
When there is a wrong value of POS which is not in the tagset, we need ignore this sentence. The "flag" is used to control this. Maybe I should use a more specific name.

reckart reviewed (commented):
The code uses many magic strings and some magic numbers. It would be great to have constants for them which also explain what they mean and what they are used for.

In general, the code could profit from line comments explaining what is going on, in particular in the more algorithmic parts like the BIO handling or when flair is called explaining why these particular hyper-parameters were chosen.

bjtulinl committed:
#19 - Refactor the code and add tests

- Add comments to the functions
- Update the BIO encoding

jcklie commented on app_flair.py:
```diff
@@ -68,18 +81,22 @@ def route_predict_ner():
+        t1 = threading.Thread(target=flair_train_ner, args=(train, dev))
         t1.start()
-        return ('', 204)
+        return "Successful training", 204
```
You can use `http.HTTPStatus`, e.g by returning ` HTTPStatus.TOO_MANY_REQUESTS`. Same for the other status codes.

jcklie commented on app_flair.py:
```diff
@@ -68,18 +81,22 @@ def route_predict_ner():
+        return "Successful training", 204
     else:
-        return ('', 429)
+        return "Request refused. Too many training requests have been sent.", 429
```
It should mean that training is still in progress

jcklie commented on app_flair.py:
```diff
@@ -113,129 +134,227 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+        s = Sentence()
+        s.tokens = tokens
+        list_sentences.append(s)
+    global test_sentences
```
Why do we need this global?

jcklie commented on app_flair.py:
```diff
@@ -113,129 +134,227 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+        token.add_tag("ner", "O")
+        i += 1
+        tokens.append(token)
+    return False, tokens
```
Why return false?

jcklie commented on app_flair.py:
```diff
@@ -113,129 +134,227 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+    return False, tokens
+
+
+def flair_train_ner(train: list, dev: list):
```
In python typing, you should use the generic `List` type, see e.g. https://docs.python.org/3/library/typing.html#typing.List

jcklie commented on app_flair.py:
```diff
@@ -51,13 +61,16 @@
 
 global tagger_pos
```
For the globals, you could use an object on which you set and get the networks. That looks much better imho than global variables.

jcklie commented on app_flair.py:
```diff
@@ -113,129 +134,227 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+        cas = load_cas_from_xmi(document["xmi"], typesystem=typesystem)
+        list_from_document = flair_train_pos_dataset(cas)
+        list_sentences.extend(list_from_document)
+    print(len(list_sentences))
```
Is this still needed?

jcklie commented on app_flair.py:
```diff
@@ -113,129 +134,227 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
-        tokens = [Token(cas.get_covered_text(t)) for t in list(cas.select_covered(TOKEN_TYPE, sentence))]
+        tokens = [
+            Token(cas.get_covered_text(t))
+            for t in list(cas.select_covered(TOKEN_TYPE, sentence))
```
No list needed I think

jcklie commented on app_flair.py:
```diff
@@ -268,13 +387,22 @@ def predict_pos(prediction_request: PredictionRequest) -> PredictionResponse:
-        tokens = [Token(cas.get_covered_text(t)) for t in list(cas.select_covered(TOKEN_TYPE, sentence))]
+        tokens = [
+            Token(cas.get_covered_text(t))
+            for t in list(cas.select_covered(TOKEN_TYPE, sentence))
```
No list needed I think

jcklie commented on app_flair.py:
```diff
@@ -289,44 +417,63 @@ def predict_pos(prediction_request: PredictionRequest) -> PredictionResponse:
+    return cas
+
+
+class TestBIOEncoding(unittest.TestCase):
```
Please move this to a new folder "tests" and put it into an extra file. Add in the readme how to run the tests.

jcklie replied:
I ran it with `python -m unittest discover` from the command line

bjtulinl replied:
I ran the test dynamically. I uploaded a document with the correct labels in INCEpTION and sent a request. After BIO encoding the document, I use the unittest to compare it with the original document. In this case, I think the test can not be run directly from the command line.

jcklie replied:
The test needs to be runnable standalone though in an isolated setting.

jcklie reviewed (changes requested):
It looks ok, but needs some improvements. Especially correct typing, removing globals and putting tests into an extra file is needed.

jcklie commented:
@bjtulinl Can you resolve this conflict?

jcklie commented:
@bjtulinl I think we first need to wrap up this first.

bjtulinl committed:
#19 - Refactor the code and add tests
- Refactor the test case into a new folder
- Fix the bug caused by too small dataset
- The file "test_n.txt" is used for testing the BIO encoding.

jcklie reviewed (changes requested):
- The tests fails for me
- There is a conflict that needs to be resolved before merging
- There are some minor remarks which are not addressed yet

You are almost there!

bjtulinl committed:
#19 - Refactor the code and add tests
- Modify the test case

bjtulinl committed:
Merge branch 'master' into feature/19-Refactor-the-code-and-add-tests

jcklie commented on tests/test_bio_encoding.py:
```diff
@@ -0,0 +1,92 @@
+NER_TYPE = "de.tudarmstadt.ukp.dkpro.core.api.ner.type.NamedEntity"
+
+
+def bio_encoding(cas, sentence, tagset, t):
```
Why is this code here? It should not be in the test folder...

bjtulinl replied:
When I import the bio_encoding function from app_flair, it also needs to load the models and get the model parameters from the config file of gunicorn. But if we only run the test case, it will not load the config file of gunicorn.

jcklie replied:
This should be prevented by the `if __name__ == "__main__":`. The else [here](https://github.com/inception-project/external-recommender-flair/blob/7093b35f04bd86bbe928451fb96cf94b8faa9684/app_flair.py#L464) looks strange, is it causing that?

bjtulinl replied:
Yes. The "else" part is used when we start the application with gunicorn and load the parameters from the config file.

jcklie replied:
Can you check whether using an `elif` with the solution in [here](https://stackoverflow.com/questions/22192519/detect-if-flask-is-being-run-via-gunicorn/25877937) works?

bjtulinl replied:
Thanks. I will check it.

bjtulinl committed:
#19 - Refactor the code and add tests
- Modify the test case

bjtulinl committed:
Merge remote-tracking branch 'origin/feature/19-Refactor-the-code-and-add-tests' into feature/19-Refactor-the-code-and-add-tests

jcklie committed:
#19 - Refactor the code and add tests

- Update README

jcklie committed:
Merge branch 'feature/19-Refactor-the-code-and-add-tests' of github.com:inception-project/external-recommender-flair into feature/19-Refactor-the-code-and-add-tests
